cli/flox-rust-sdk/src/providers/publish.rs

```diff
@@ -0,0 +1,90 @@
+use crate::flox::Flox;
+use std::path::Path;
+
+pub struct PublishProvider {
+    repo_dir: Path,
+    package_name: String,
+}
+
+impl PublishProvider {
+    pub fn new(repo_dir: Path, package: &str) -> Self {
+        Self { repo_dir: repo_dir, package_name: package.to_string() }
+    }
+
+    /// initiates the publish process
+    pub fn publish(&self) -> Result<(), String> {
+        // Run each phase, gathering info along the way until we get to the end
+        // and can send it all to the catalog
+
+        // Best pattern to use here?
```

Comment:
My preference would be bare functions that take typestate-like arguments, something like this:
```rust
pub struct CheckedMetadata {
    pub git_metadata: GitMetadata,
    pub base_catalog_metadata: BaseCatalogMetadata,
    pub package_metadata: PackageMetadata,
    pub user_catalog_metadata: UserCatalogMetadata,
    pub repo_dir: PathBuf,
    pub package_name: String,
    // This field isn't "pub", so no one outside this module can construct this struct. That helps
    // ensure that we can only make this struct as a result of doing the "right thing."
    _private: ()
}

pub struct VerifiedPackageBuild {
    git_metadata: GitMetadata,
    base_catalog_metadata: BaseCatalogMetadata,
    package_metadata: PackageMetadata,
    user_catalog_metadata: UserCatalogMetadata,
    repo_dir: PathBuf,
    package_name: String,
    sandbox_dir: PathBuf, // <- new field since build has a side effect
    _private: ()
}

pub fn check_metadata(repo_dir: impl AsRef<Path>, package_name: impl AsRef<str>) -> Result<CheckedMetadata, Error> {
    let git_metadata = gather_git_metadata(repo_path);
    let base_catalog_metadata = gather_base_catalog_metadata(lockfile_path);
    let package_metadata = gather_publishable_package_metadata(lockfile_path);
    let user_catalog_metadata = gather_user_catalog_metadata(???);
    if metadata_is_good() {
        CheckedMetadata {
            git_metadata,
            base_catalog_metadata,
            package_metadata,
            user_catalog_metadata,
            repo_dir: repo_dir.as_ref().into_path_buf(),
            package_name: package_name.as_ref().to_string()
        }
    }
}

/// Impossible to check the build if the metadata hasn't already been checked
pub fn check_build<B: ManifestBuilder>(metadata: CheckedMetadata, builder: &B, sandbox: impl AsRef<Path>) -> Result<VerifiedPackageBuild, Error> {
    ...
}

// ...rest of the publish process
```

I like this because it keeps related data bundled together in the structs, but the operations are pure. In addition, the structs can only be constructed by the functions provided in this module, so a consumer shouldn't be able to pick up in the middle of the publish process by mistake (tests in this module will still be able to though).

An alternative is to put these pure functions on `PublishProvider` as static methods. I've never quite understood the motivation for having a bunch of static functions associated with a struct, seems like free standing functions do the same thing with less typing.

Reply:
That looks good and I think is similar to what @ysndr suggested.